README.md
# React Carousel

A React carousel component that aims to be as flexible as possible.

Installation
---

```sh
npm i -S @moxy/react-carousel
```

or

```sh
yarn add @moxy/react-carousel
```

This library is written in modern JavaScript and is published in both CommonJS
and ES module transpiled variants. If you target older browsers please make
sure to transpile accordingly.

Motivation
---

After analysing and testing a number of different carousel solutions for react,
[MOXY](https://moxy.studio) determined that none of the existing solutions
would be a good candidate the numerous projects that we develop as an agency.
As such, we set out to build an extensible, easy to use, small carousel
component.

This project is still a work in progress. You can see what is currently missing
in [Future Work](#future-work).

### How does `@moxy/react-carousel` differ from other packages?

- Uses the minimum amount of dependencies.
    - The only direct dependency is `lodash.debounce`.
- Does not highjack native scrolling functionality unless desired.
- Respects native browser scrolling behaviour (desktop and mobile), and only aims to augment its capabilities.
    - Once Safari fixes its issues with `scroll-snap` and programmatic manipulation of `scrollLeft` / `scrollTop`, js-based snapping behaviour can be dropped in favor of `scroll-snap`.
- Manipulates carousel state by changing `scrollLeft` property.
- Does not use inline styles for controlling behaviour (except [`offset`](#offset)); all behaviour is controlled via className exchange and CSS rules. All classNames can be customized.
- Developer / designer oriented: `@moxy/react-carousel` does not assume or enforce a particular style of carousel, and provides a lot of options for customizing its behaviour.
- Provides callbacks before and after current slide changes

Usage
---

`@moxy/react-carousel` does not make any assumptions on what elements your
slides represent. They can be a `<div>`, an `<img>`, or any other element. They
can also be react components, but you must be aware that the `key`,
`onMouseUp`, `onDragStart`, and `onDrag` will be overriden. `className` prop
will be appended to `@moxy/react-carousel`'s classNames.

### Basic usage

```js
import Carousel from '@moxy/react-carousel';

<Carousel>
    <div>Slide 1</div>
    <div>Slide 2</div>
    <div className="custom">Slide 3</div>
    <div>Slide 4</div>
</Carousel>
```

### Custom arrows rendering

```js
import Carousel from '@moxy/react-carousel';

<Carousel
    arrows
    renderArrows={ ({ previous, next }) => (
        <>
            <span onClick={ previous } />
            <span onClick={ next }>
        </>
    )}>
    <div>Slide 1</div>
    <div>Slide 2</div>
    <div>Slide 3</div>
    <div>Slide 4</div>
</Carousel>
```

There is a lot more than you can do, check out the
[demo](https://moxystudio.github.io/react-carousel), where you can easily
explore different prop values and prop combinations.

HTML Structure
---

```js
<Carousel arrows dots>
    <div>Slide 1</div>
    <img src={ slide2 } alt="Slide 2" />
    <div className="custom">Slide 3</div>
</Carousel>
```

will render:

```js
<div className="rc-wrapper">
    <div className="rc">
        <div className="rc-slider">
            <div className="rc-slide -current">Slide </div>
            <img className="rc-slide" src={ slide2 } alt="Slide 2" />
            <div class="rc-slide custom">Slide 3<div>
        </div>
    </div>

    <button className="rc-arrow -left">previous</button>
    <button className="rc-arrow -right">next</button>

    <div className="rc-dots">
        <button className="rc-dot -current" />
        <button className="rc-dot" />
        <button className="rc-dot" />
    </div>
</div>
```

CSS
---

There isn't one definitive implementation suitable for `@moxy/react-carousel`.

A [default stylesheet](src/styles.css) is shipped with the package, and can
be imported with:

```
import `@moxy/react-carousel/dist/styles.css`
```

You can use this stylesheet as is and build on top of it, or use it as
inspiration for your own styles. The default styles sets `.rc-slide` to
`display: inline-block`, and, `.rc-slider` to `white-space: nowrap` so that
slides are positioned horizontally. You could achieve the same effects with
`float`, `flexbox`, etc. Because the position for the current slide is
calculated based on the distance from the element to its parent, it also allows
for different size slides

Props
---

| **Prop**                                                          | **Type**               | **Required** | **Default**   |
| ---                                                               | ---                    | ---          | ---           |
| [`children`](#children)                                           | `any`                  | **yes**      | `undefined`   |
| [`arrows`](#arrows)                                               | `boolean`              | no           | `false`       |
| [`dots`](#dots)                                                   | `boolean`              | no           | `false`       |
| [`disableNativeScroll`](#disableNativeScroll)                     | `boolean`              | no           | `false`       |
| [`draggable`](#draggable)                                         | `boolean`              | no           | `false`       |
| [`infinite`](#infinite)                                           | `boolean`              | no           | `false`       |
| [`keyboardControl`](#keyboardControl)                             | `boolean`              | no           | `false`       |
| [`resetCurrentOnResize`](#resetCurrentOnResize)                   | `boolean`              | no           | `true`        |
| [`swapOnDragMoveEnd`](#swapOnDragMoveEnd)                         | `boolean`              | no           | `true`        |
| [`autoplayIntervalMs`](#autoplayIntervalMs)                       | `number`               | no           | `0`           |
| [`autoplayDirection`](#autoplayDirection)                         | `string`               | no           | `'ltr'`       |
| [`offset`](#offset)                                               | `number` or `function` | no           | `0`           |
| [`slideSnapEasing`](#slideSnapEasing)                             | `string` or `function` | no           | `ease-in-out` |
| [`slideSnapDuration`](#slideSnapDuration)                         | `number`               | no           | `150`         |
| [`slideTransitionEasing`](#slideTransitionEasing)                 | `string` or `function` | no           | `ease-in-out` |
| [`slideTransitionDuration`](#slideTransitionDuration)             | `number`               | no           | `300`         |
| [`touchSwipeVelocityThreshold`](#touchSwipeVelocityThreshold)     | `number`               | no           | `0.3`         |
| [`touchCrossAxisScrollThreshold`](#touchCrossAxisScrollThreshold) | `number`               | no           | `0.45`        |
| [`current`](#current)                                             | `number`               | no           | `undefined`   |
| [`beforeChange`](#beforeChange)                                   | `function`             | no           | `() => {}`    |
| [`afterChange`](#afterChange)                                     | `function`             | no           | `() => {}`    |
| [`renderArrows`](#renderArrows)                                   | `function`             | no           | `undefined`   |
| [`renderDots`](#renderDots)                                       | `function`             | no           | `undefined`   |
| [`wrapperClassName`](#wrapperClassName)                           | `string`               | no           | `undefined`   |
| [`carouselClassName`](#carouselClassName)                         | `string`               | no           | `undefined`   |
| [`sliderClassName`](#sliderClassName)                             | `string`               | no           | `undefined`   |
| [`arrowClassName`](#arrowClassName)                               | `string`               | no           | `undefined`   |
| [`dotContainerClassName`](#dotContainerClassName)                 | `string`               | no           | `undefined`   |
| [`dotClassName`](#dotClassName)                                   | `string`               | no           | `undefined`   |
| [`modifierDraggableClassName`](#modifierDraggableClassName)       | `string`               | no           | `undefined`   |
| [`modifierDraggingClassName`](#modifierDraggingClassName)         | `string`               | no           | `undefined`   |
| [`modifierCurrentClassName`](#modifierCurrentClassName)           | `string`               | no           | `undefined`   |
| [`modifierLeftClassName`](#modifierLeftClassName)                 | `string`               | no           | `undefined`   |
| [`modifierRightClassName`](#modifierRightClassName)               | `string`               | no           | `undefined`   |

API
---

#### `children`

Type: `any`

Default: `undefined`

`@moxy/react-carousel` assumes that any direct descendants of the `<Carousel>`
component are the slides that you wish to render. It makes no assumptions on
what those elements should be, or their dimensions.

#### `arrows`

Type: `boolean`

Default: `false`

Render previous / next buttons to navigate carousel. Default arrows are
rendered as `<button>` elements. `arrowClassName` and `modifierLeftClassName`
applied to the 'Previous' button, and `arrowClassName` and
`modifierRightClassName` are applied to 'Next' button.

#### `dots`

Type: `boolean`

Default: `false`

Render one button per slide to navigate / indicate carousel state. Default dots
are rendered as `<button>` elements. `dotClassName` is applied, as well as
`modifierCurrentClassName` to the button that matches the current slide.

#### `disableNativeScroll`

Type: `boolean`

Default: `false`

It `true`, `modifierDisableScrollClassName` is applied to `.rc` element

#### `draggable`

Type: `boolean`

Default: `false`

If `true`, enables grabbing and dragging the slider with the mouse to change
its scroll value.

NOTE: When `false`, it only prevents dragging with mouse, it does not prevent
dragging with touch events in devices that support it

#### `infinite`

Type: `boolean`

Default: `false`

If `true`, the carousel will navigate to the first / last slide if attempting
to navigate out of bounds

#### `keyboardControl`

Type: `boolean`

Default: `false`

If `true`, enables focusing the `.rc` element. While focused, you are able to
switch slides using the left and right arrow keys.

#### `resetCurrentOnResize`

Type: `boolean`

Default: `true`

If `true`, carousel will re-render with the current state after the window has
been resized. This operation is debounced by 200ms.

#### `swapOnDragMoveEnd`

Type: `boolean`

Default: `true`

If `true`, the carousel will change the current slide, to an adjacent one,
based on the direction of the drag movement.
Although, if the drag is enough to change the current slide by itself, it will
respect that behaviour.

*Notes: Only works if the `draggable` prop is set as `true`.*

#### `autoplayIntervalMs`

Type: `number`

Default: `0`

The number of milliseconds to wait between automatic slide advancement. A value
of `0` turns off autoplay.

#### `autoplayDirection`

Type: `oneOf(['ltr', 'rtl'])`

Default: `'ltr'`

The direction that the carousel should change the slides when
`autoplayIntervalMs > 0`. Can be one of:

- `'ltr'`: causes carousel to change slides from left to right
- `'rtl'`: causes carousel to change slides from right to left

#### `offset`

Type: `number` or `function`

Default: `0`

The horizontal padding given to `.rc-slider`.

If it is a number, is treated as a value in pixels.

If it is a function, its signature is:

```
({ animating: boolean, dragging: boolean, current: number }) => number
```

It is expected that the returned number represents a value in pixels.

#### `slideSnapEasing`

Type: `function` or `oneOf(['linear', 'ease-in', 'ease-out', 'ease-in-out'])`

Default: `ease-in-out`

Easing function applied to snapping animation.

You may instead provide your own easing function if the supplied ones are not
suitable.  `slideSnapEasing` also accepts a function, `(value: number) =>
number`. Return value must be in the `[0.0, 1.0]` range.

#### `slideSnapDuration`

Type: `number`

Default: `150`

Number of milliseconds that the snapping animation takes to complete.

#### `slideTransitionEasing`

Type: `function` or `oneOf(['linear', 'ease-in', 'ease-out', 'ease-in-out'])`

Default: `ease-in-out`

Easing function applied to transition animation.

You may instead provide your own easing function if the supplied ones are not
suitable.  `slideTransitionEasing` also accepts a function, `(value: number) =>
number`. Return value must be in the `[0.0, 1.0]` range.

#### `slideTransitionDuration`

Type: `number`

Default: `300`

Number of milliseconds that the transition animation takes to complete.

#### `touchSwipeVelocityThreshold`

Type: `number`

Default: `0.3`

**Note**: Only applicable when `disableNativeScroll = true`.

The velocity in pixels per frame where a swipe causes the slide to change

#### `touchCrossAxisScrollThreshold`

Type: `number`

Default: `0.45`

**Note**: Only applicable when `disableNativeScroll = true`.

The velocity in pixels per frame in the cross axis (currently just y axis)
below where a diagonal swipe on the carousel will prevent vertical scrolling

#### `current`

Type: `number`

Default: `undefined`

If set, makes the Carousel as a [controlled
component](https://reactjs.org/docs/forms.html#controlled-components), allowing
you to control the current slide from outside the scope of the Carousel. This
will not disable any internal state management, and `beforeChange` /
`afterChange` props will still be called, so you will be responsible for
syncing your external state with the internal one of the Carousel.

#### `beforeChange`

Type: `function`

Default: `() => {}`

Callback function that is called before a slide change happens.

Its signature is `({ current: number, next: number, source: 'user' | 'autoplay' }) => void`

#### `afterChange`

Type: `function`

Default: `() => {}`

Callback function that is called after a slide change happens. Function is only
called after transitions have ended and internal state is updated.

Its signature is `({ previous: number, current: number, source: 'user' | 'autoplay' }) => void`

#### `renderArrows`

Type: `(props: object) => React.Node`

Default: `undefined`

If a function is provided to `renderArrows` prop, the default arrows will not
be rendered, and you may use this render prop to [render your own
arrows](#custom-arrows-rendering).

`props` has the following shape:

```js
{
    next: () => void,      // call this to change to the next (right) slide
    previous: () => void,  // call this to change to the previous (left) slide
    current: number,       // index of the current slide
    animating: boolean,    // boolean indicating that a transition or snap is happening
    dragging: boolean,     // boolean indicating that the carousel is being dragged by a mouse
    slideCount: number,    // number indicating how many slides are rendered by the carousel
}
```

#### `renderDots`

Type: `(props: object) => React.Node`

Default: `undefined`

If a function is provided to `renderDots` prop, the default dots will not
be rendered, and you may use this render prop to render your own dots.

`props` has the following shape:

```js
{
    setCurrent: (i: number) => void,  // call this to change to slide with index i
    current: number,                  // index of the current slide
    animating: boolean,               // boolean indicating that a transition or snap is happening
    dragging: boolean,                // boolean indicating that the carousel is being dragged by a mouse
}
```

### Appending element classNames

You may append the default classNames rendered by `<Carousel>` with your own classNames. Each type of
element has its own className, and some of them also have modifier classNames
applied.

#### `wrapperClassName`

Type: `string`

Default: `undefined`

Applied to top level element rendered by `<Carousel>`, `.rc-wrapper`. It will have
`.rc`, `.rc-arrow` and `.rc-dots` elements as its direct descendants.

#### `carouselClassName`

Type: `string`

Default: `undefined`

Applied to scrollable element.

Expected to have at least `overflow: x; -webkit-overflow-scrolling: touch` applied.

#### `sliderClassName`

Type: `string`

Default: `undefined`

Applied to slides container element. This element can be larger than the full
width of the viewport, and is scrolled by its direct parent, the `.rc` element

#### `arrowClassName`

Type: `string`

Default: `undefined`

Applied to "previous" and "next" `<button>` elements rendered by default when
`arrows` is `true` and no custom `renderArrows` prop is provided.

#### `dotContainerClassName`

Type: `string`

Default: `undefined`

Applied to container `<div>` element that holds the dots rendered by default
when `dots` is `true` and no custom `renderDots` prop is provided.

#### `dotClassName`

Type: `string`

Default: `undefined`

Applied to each individual `<button>` dot element rendered by default when
`dots` is `true` and no custom `renderDots` prop is provided.

### Appending modifier classNames

#### `modifierDraggableClassName`

Type: `string`

Default: `undefined`

Appended to `.rc-slider.-draggable` when `draggable` prop is true.

#### `modifierDraggingClassName`

Type: `string`

Default: `undefined`

Appended to `.rc-slider.-draggable.-dragging` when carousel is being dragged by
mouse. Useful to change icons.

#### `modifierCurrentClassName`

Type: `string`

Default: `undefined`

Appended to `.rc-slide.-current` and `.rc-dot.-current` elements. Represents
the currently selected slide / dot.

#### `modifierLeftClassName`

Type: `string`

Default: `undefined`

Appended to `.rc-arrow.-left` that switches to the previous (left) slide.

#### `modifierRightClassName`

Type: `string`

Default: `undefined`

Appended to `.rc-arrow.-right` that switches to the next (right) slide

Future work
---

- [ ] Infinite mode
    - [x] enable / disable
    - [x] bounce
    - [ ] wrap around
- [ ] Current slide placement options
    - [x] left
    - [ ] center
    - [ ] right
- [ ] Vertical carousel and vertical placement options
    - [ ] top
    - [ ] center
    - [ ] bottom


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.10](https://github.com/moxystudio/react-carousel/compare/v0.1.9...v0.1.10) (2020-11-04)


### Bug Fixes

* fix click behaviour ([bf2dcda](https://github.com/moxystudio/react-carousel/commit/bf2dcdae31296647c9d19ffb4db44201f21baf88))

### [0.1.9](https://github.com/moxystudio/react-carousel/compare/v0.1.8...v0.1.9) (2020-08-19)


### Features

* support conditional / null carousel children ([c244329](https://github.com/moxystudio/react-carousel/commit/c244329f068f561553cca9e15239037a6459eac0))

### [0.1.8](https://github.com/moxystudio/react-carousel/compare/v0.1.7...v0.1.8) (2020-06-09)


### Features

* add source parameter to afterChange and beforeChange callbacks ([5346752](https://github.com/moxystudio/react-carousel/commit/53467528fedead5b5b5934be6e6f298a5be949b5))


### Bug Fixes

* reset autoplay when current changes ([53a57c2](https://github.com/moxystudio/react-carousel/commit/53a57c280b2a8eecc4733b4169cae8eeb6506030))

### [0.1.7](https://github.com/moxystudio/react-carousel/compare/v0.1.6...v0.1.7) (2020-05-28)


### Bug Fixes

* reset autoplay interval on slide click ([9935ac1](https://github.com/moxystudio/react-carousel/commit/9935ac12870d05cab45babf1d78b8a37a959b315))

### [0.1.6](https://github.com/moxystudio/react-carousel/compare/v0.1.5...v0.1.6) (2020-04-22)


### Features

* allow using functions or strings in easing props ([1fbee7f](https://github.com/moxystudio/react-carousel/commit/1fbee7fdadfbe0f5854dabdd9056cd03e1475639))

### [0.1.5](https://github.com/moxystudio/react-carousel/compare/v0.1.4...v0.1.5) (2020-04-13)


### Features

* add slidecount arg to renderarrows ([8575fbf](https://github.com/moxystudio/react-carousel/commit/8575fbf5fb36cafd240a36ddaf9f941795ebc78b))
* controlled carousel ([fa2fe89](https://github.com/moxystudio/react-carousel/commit/fa2fe8947d7b3dcf8b11b983418d93c50e3a5bd1))


### Bug Fixes

* enable default arrows in infinite mode ([a143c9f](https://github.com/moxystudio/react-carousel/commit/a143c9f0e092757bb00a7a13f37be9fb81f6bdba))
* scroll to origin on mount ([8c201d5](https://github.com/moxystudio/react-carousel/commit/8c201d50afe46442faef76e537ebd45ef25b56f6))

### [0.1.4](https://github.com/moxystudio/react-carousel/compare/v0.1.3...v0.1.4) (2020-04-03)


### Features

* add cross-env to support setting env variables in multiple platforms ([13a8e11](https://github.com/moxystudio/react-carousel/commit/13a8e11f39cae7a12ccd557d159dee55175ae382))
* swap slides based on drag movement ([6001ae5](https://github.com/moxystudio/react-carousel/commit/6001ae54929267c59a533559b3d7d5a1f0732ddc))

### [0.1.3](https://github.com/moxystudio/react-carousel/compare/v0.1.2...v0.1.3) (2020-04-01)


### Bug Fixes

* allow scroll when touch movement is vertical ([bfb3293](https://github.com/moxystudio/react-carousel/commit/bfb3293bdc295e9991a20005ddc04e5552a8e6bd))

### [0.1.2](https://github.com/moxystudio/react-carousel/compare/v0.1.1...v0.1.2) (2020-03-25)


### Bug Fixes

* cancel debounced handlers on component unmount ([3331112](https://github.com/moxystudio/react-carousel/commit/333111241e281b01f8efcb73d8310051869b1d6d))

### [0.1.1](https://github.com/moxystudio/react-carousel/compare/v0.1.0...v0.1.1) (2020-03-10)


### Bug Fixes

* include `dist` in package.json `files` ([#4](https://github.com/moxystudio/react-carousel/issues/4)) ([4506471](https://github.com/moxystudio/react-carousel/commit/450647131a714aa58ac6ac5b89a9c4df19b7006e))
* **docs:** readme typos fixed ([b31242b](https://github.com/moxystudio/react-carousel/commit/b31242bfb2f97727ed8a122194b9895ca6e99541))

## 0.1.0 (2020-03-10)


### Features

* demo application ([9ca073d](https://github.com/moxystudio/react-carousel/commit/9ca073d9802b0e9b152a69138b7bc7d6576e81ed))
* initial implementation ([e790885](https://github.com/moxystudio/react-carousel/commit/e790885c305683020b39a5486328bcb7a7fd0560))


demo/pages/index.js
import React, { useState, useReducer } from 'react';
import Carousel from '@moxy/react-carousel';
import { isArray } from 'lodash';
import classnames from 'classnames';

import styles from './index.module.css';

const renderSlides = (length = 10) => (
    Array
        .from({ length })
        .map((_, i) => (
            <img
                key={ i }
                src={ `https://placehold.it/300x300/${i}${i}${i}/fff` } />
        ))
);

const carouselProps = [
    [
        ['arrows', 'checkbox'],
        ['dots', 'checkbox'],
        ['disableNativeScroll', 'checkbox'],
        ['draggable', 'checkbox'],
        ['infinite', 'checkbox'],
        ['keyboardControl', 'checkbox'],
        ['resetCurrentOnResize', 'checkbox'],
        ['swapOnDragMoveEnd', 'checkbox'],
    ],
    [
        ['autoplayIntervalMs', 'number'],
        ['autoplayDirection', ['ltr', 'rtl']],
        ['offset', 'number'],
        ['slideSnapDuration', 'number'],
        ['slideSnapEasing', ['linear', 'ease-in', 'ease-out', 'ease-in-out']],
        ['slideTransitionDuration', 'number'],
        ['slideTransitionEasing', ['linear', 'ease-in', 'ease-out', 'ease-in-out']],
        ['touchSwipeVelocityThreshold', 'number'],
        ['touchCrossAxisScrollThreshold', 'number'],
        ['current', 'number'],
    ],
    [
        ['wrapperClassName', 'text'],
        ['carouselClassName', 'text'],
        ['sliderClassName', 'text'],
        ['arrowClassName', 'text'],
        ['dotContainerClassName', 'text'],
        ['dotClassName', 'text'],
        ['modifierDraggableClassName', 'text'],
        ['modifierDraggingClassName', 'text'],
        ['modifierCurrentClassName', 'text'],
        ['modifierLeftClassName', 'text'],
        ['modifierRightClassName', 'text'],
    ],
];

const renderInput = (prop, type, state, setState) => {
    const attr = type === 'checkbox' ? 'checked' : 'value';

    return (
        <label key={ prop } className={ styles.propSetterLabel }>
            {isArray(type) && (
                <select
                    className={ styles.propSetterInput }
                    onChange={ (ev) => setState({ [prop]: ev.target[attr] }) }
                    value={ state[prop] }>
                    {type.map((option) => (
                        <option key={ option } value={ option }>{ option }</option>
                    ))}
                </select>
            )}

            {!isArray(type) && (
                <input
                    className={ type === 'checkbox' ? styles.propSetterCheckbox : styles.propSetterInput }
                    { ...{
                        type,
                        [attr]: state[prop],
                        onChange: (ev) => setState({
                            [prop]: (
                                type === 'number' ? parseFloat(ev.target[attr]) : ev.target[attr]
                            ),
                        }),
                    } } />
            )}

            <span>{prop}</span>
        </label>
    );
};

// eslint-disable-next-line react/prop-types
const renderArrows = ({ previous, next }) => (
    <>
        <span className={ `${styles.customArrow} ${styles.customArrowLeft}` } onClick={ previous }>←</span>
        <span className={ `${styles.customArrow} ${styles.customArrowRight}` } onClick={ next }>→</span>
    </>
);

// eslint-disable-next-line react/prop-types
const renderDots = ({ setCurrent, current, slideCount }) => (
    <ul className="rc-dots">
        {Array.from({ length: slideCount }).map((_, i) => (
            <li
                className={ classnames(styles.customDot, { [styles.customDotCurrent]: current === i }) }
                key={ i }
                onClick={ () => setCurrent(i) } />
        ))}
    </ul>
);

// eslint-disable-next-line react/prop-types
const PropSetter = ({ children }) => {
    const [logCallbacks, setLogCallbacks] = useState(false);
    const [customRenderArrows, setCustomRenderArrows] = useState(false);
    const [customRenderDots, setCustomRenderDots] = useState(false);

    const [state, setState] = useReducer(
        (state, action) => ({ ...state, ...action }),
        {
            arrows: false,
            dots: false,
            disableNativeScroll: false,
            draggable: false,
            infinite: false,
            keyboardControl: false,
            resetCurrentOnResize: true,
            swapOnDragMoveEnd: true,
            autoplayIntervalMs: 0,
            autoplayDirection: 'ltr',
            slideSnapDuration: 150,
            slideSnapEasing: 'ease-in-out',
            slideTransitionDuration: 300,
            slideTransitionEasing: 'ease-in-out',
            touchSwipeVelocityThreshold: 0.3,
            touchCrossAxisScrollThreshold: 0.45,
            offset: 100,
            current: 0,
            renderArrows: undefined,
            renderDots: undefined,
            wrapperClassName: undefined,
            carouselClassName: undefined,
            sliderClassName: undefined,
            arrowClassName: undefined,
            dotContainerClassName: undefined,
            dotClassName: undefined,
            modifierDraggableClassName: undefined,
            modifierDraggingClassName: undefined,
            modifierCurrentClassName: undefined,
            modifierLeftClassName: undefined,
            modifierRightClassName: undefined,
            modifierDisableScrollClassName: undefined,
        },
    );

    const callbacks = {
        beforeChange: logCallbacks ? (v) => console.log('beforeChange', v) : () => {},
        afterChange: logCallbacks ? (v) => console.log('afterChange', v) : () => {},
        onUserInteraction: logCallbacks ? (type, ev) => console.log('onUserInteraction', type, ev) : () => {},
    };

    const renderProps = {
        renderArrows: customRenderArrows ? renderArrows : undefined,
        renderDots: customRenderDots ? renderDots : undefined,
    };

    return (
        <>
            <div className={ styles.propSetter }>
                {carouselProps.map((column, i) => (
                    <div key={ i } className={ styles.propSetterColumn }>
                        {column.map((item) => renderInput(...item, state, setState))}
                    </div>
                ))}
            </div>

            <div className={ styles.propSetter }>
                <div className={ styles.propSetterColumn }>
                    <label>
                        <input
                            type="checkbox"
                            checked={ customRenderArrows }
                            onChange={ (ev) => setCustomRenderArrows(ev.target.checked) } />
                        Use custom renderArrows
                    </label>

                    <label>
                        <input
                            type="checkbox"
                            checked={ customRenderDots }
                            onChange={ (ev) => setCustomRenderDots(ev.target.checked) } />
                        Use custom renderDots
                    </label>
                </div>

                <div className={ styles.propSetterColumn }>
                    <label>
                        <input
                            type="checkbox"
                            checked={ logCallbacks }
                            onChange={ (ev) => setLogCallbacks(ev.target.checked) } />
                        Log beforeChange and afterChange to console
                    </label>
                </div>
            </div>

            {children({ ...state, ...callbacks, ...renderProps })}
        </>
    );
};

const Home = () => (
    <div className={ styles.home }>
        <h1 className={ styles.title }>@moxy/react-carousel explorer</h1>

        <PropSetter>
            {(props) => (
                <Carousel { ...props }>
                    {renderSlides()}
                </Carousel>
            )}
        </PropSetter>
    </div>
);

export default Home;
